// typhonjs-escomplex-commons/dist/utils/Enum.js
'use strict';

Object.defineProperty(exports, "__esModule", {
   value: true
});

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * This is an abstract class that is not intended to be used directly. Extend it to turn your class into an enum
 * (initialization is performed via `MyClass.initEnum()`).
 *
 * Modified from source provided by enumify (license unlisted / public domain)
 * @see https://github.com/rauschma/enumify
 */
var Enum = function () {
   /**
    * `initEnum()` closes the class. Then calling this constructor throws an exception.
    *
    * If your subclass has a constructor then you can control what properties are added to `this` via the argument you
    * pass to `super()`. No arguments are fine, too.
    *
    * @param {object}   instanceProperties - Provides initial properties.
    */
   function Enum() {
      var instanceProperties = arguments.length <= 0 || arguments[0] === undefined ? void 0 : arguments[0];

      _classCallCheck(this, Enum);

      // new.target would be better than this.constructor, but isn’t supported by Babel
      if ({}.hasOwnProperty.call(this.constructor, INITIALIZED)) {
         throw new Error('Enum classes can’t be instantiated');
      }

      if ((typeof instanceProperties === 'undefined' ? 'undefined' : _typeof(instanceProperties)) === 'object' && instanceProperties !== null) {
         s_COPY_PROPERTIES(this, instanceProperties);
      }
   }

   /**
    * Set up the enum, close the class.
    *
    * @param {Array|object}   arg - Either an object whose properties provide the names and values (which must be
    * mutable objects) of the enum constants. Or an Array whose elements are used as the names of the enum constants.
    * The values are create by instantiating the current class.
    *
    * @returns {Enum}
    */


   _createClass(Enum, [{
      key: 'toString',


      /**
       * Default `toString()` method for enum constant.
       *
       * @returns {string}
       */
      value: function toString() {
         return this.constructor.name + '.' + this.name;
      }
   }], [{
      key: 'initEnum',
      value: function initEnum(arg) {
         Object.defineProperty(this, 'enumValues', {
            value: [],
            configurable: false,
            writable: false,
            enumerable: true
         });

         if (Array.isArray(arg)) {
            this._enumValuesFromArray(arg);
         } else {
            this._enumValuesFromObject(arg);
         }

         Object.freeze(this.enumValues);
         this[INITIALIZED] = true;

         return this;
      }
   }, {
      key: '_enumValuesFromArray',
      value: function _enumValuesFromArray(arr) {
         var _iteratorNormalCompletion = true;
         var _didIteratorError = false;
         var _iteratorError = undefined;

         try {
            for (var _iterator = arr[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
               var key = _step.value;
               this._pushEnumValue(new this(), key);
            }
         } catch (err) {
            _didIteratorError = true;
            _iteratorError = err;
         } finally {
            try {
               if (!_iteratorNormalCompletion && _iterator.return) {
                  _iterator.return();
               }
            } finally {
               if (_didIteratorError) {
                  throw _iteratorError;
               }
            }
         }
      }
   }, {
      key: '_enumValuesFromObject',
      value: function _enumValuesFromObject(obj) {
         var _iteratorNormalCompletion2 = true;
         var _didIteratorError2 = false;
         var _iteratorError2 = undefined;

         try {
            for (var _iterator2 = Object.keys(obj)[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
               var key = _step2.value;

               var value = new this(obj[key]);
               this._pushEnumValue(value, key);
            }
         } catch (err) {
            _didIteratorError2 = true;
            _iteratorError2 = err;
         } finally {
            try {
               if (!_iteratorNormalCompletion2 && _iterator2.return) {
                  _iterator2.return();
               }
            } finally {
               if (_didIteratorError2) {
                  throw _iteratorError2;
               }
            }
         }
      }
   }, {
      key: '_pushEnumValue',
      value: function _pushEnumValue(enumValue, name) {
         enumValue.name = name;
         enumValue.ordinal = this.enumValues.length;

         Object.defineProperty(this, name, {
            value: enumValue,
            configurable: false,
            writable: false,
            enumerable: true
         });

         this.enumValues.push(enumValue);
      }

      /**
       * Given the name of an enum constant, return its value.
       *
       * @param {string}   name - An enum name.
       *
       * @returns {T}
       */

   }, {
      key: 'enumValueOf',
      value: function enumValueOf(name) {
         return this.enumValues.find(function (x) {
            return x.name === name;
         });
      }

      /**
       * Make enum classes iterable
       *
       * @returns {Symbol.iterator}
       */

   }, {
      key: Symbol.iterator,
      value: function value() {
         return this.enumValues[Symbol.iterator]();
      }
   }]);

   return Enum;
}();

// Module private ---------------------------------------------------------------------------------------------------

/**
 * Provides a Symbol to track initialized state.
 * @type {Symbol}
 * @ignore
 */


exports.default = Enum;
var INITIALIZED = Symbol();

/**
 * Copies an objects properties.
 *
 * @param {object}   target - Target object.
 * @param {object}   source - Source object.
 *
 * @returns {object}
 * @ignore
 */
function s_COPY_PROPERTIES(target, source) {
   // Ideally, we’d use Reflect.ownKeys() here, but I don’t want to depend on a polyfill.
   var _iteratorNormalCompletion3 = true;
   var _didIteratorError3 = false;
   var _iteratorError3 = undefined;

   try {
      for (var _iterator3 = Object.getOwnPropertyNames(source)[Symbol.iterator](), _step3; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
         var key = _step3.value;

         var desc = Object.getOwnPropertyDescriptor(source, key);
         Object.defineProperty(target, key, desc);
      }
   } catch (err) {
      _didIteratorError3 = true;
      _iteratorError3 = err;
   } finally {
      try {
         if (!_iteratorNormalCompletion3 && _iterator3.return) {
            _iterator3.return();
         }
      } finally {
         if (_didIteratorError3) {
            throw _iteratorError3;
         }
      }
   }

   return target;
}
module.exports = exports['default'];
