// typhonjs-escomplex-commons/dist/utils/ast/ASTData.js
'use strict';

Object.defineProperty(exports, "__esModule", {
   value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * Defines the output data from parsing an AST tree.
 */
var ASTData = function () {
   function ASTData() {
      _classCallCheck(this, ASTData);

      this.source = '';
      this.operands = [];
      this.operators = [];
   }

   _createClass(ASTData, [{
      key: 'write',
      value: function write(string) {
         this.source += string;
      }
   }, {
      key: 'toString',
      value: function toString() {
         return this.source;
      }
   }]);

   return ASTData;
}();

exports.default = ASTData;
module.exports = exports['default'];
