// typhonjs-escomplex-commons/dist/module/report/averages/MethodAverage.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _HalsteadAverage = require('./HalsteadAverage');

var _HalsteadAverage2 = _interopRequireDefault(_HalsteadAverage);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * Provides all the averaged method metric data.
 */
var MethodAverage =
/**
 * Initializes the default averaged method data.
 */
function MethodAverage() {
  _classCallCheck(this, MethodAverage);

  /**
   * Measures the average method cyclomatic complexity of the module / class.
   * @type {number}
   */
  this.cyclomatic = 0;

  /**
   * Measures the average method cyclomatic density of the module / class.
   * @type {number}
   */
  this.cyclomaticDensity = 0;

  /**
   * Stores the averaged Halstead metric data.
   * @type {HalsteadData}
   */
  this.halstead = new _HalsteadAverage2.default();

  /**
   * Measures the average number of method parameters for the module / class.
   * @type {number}
   */
  this.params = 0;

  /**
   * Measures the average source lines of code for the module / class.
   * @type {{logical: number, physical: number}}
   */
  this.sloc = { logical: 0, physical: 0 };
};

exports.default = MethodAverage;
module.exports = exports['default'];
