// aws-sdk/clients/all.js
require('../lib/node_loader');
module.exports = {
  ACM: require('./acm'),
  APIGateway: require('./apigateway'),
  ApplicationAutoScaling: require('./applicationautoscaling'),
  AppStream: require('./appstream'),
  AutoScaling: require('./autoscaling'),
  Batch: require('./batch'),
  Budgets: require('./budgets'),
  CloudDirectory: require('./clouddirectory'),
  CloudFormation: require('./cloudformation'),
  CloudFront: require('./cloudfront'),
  CloudHSM: require('./cloudhsm'),
  CloudSearch: require('./cloudsearch'),
  CloudSearchDomain: require('./cloudsearchdomain'),
  CloudTrail: require('./cloudtrail'),
  CloudWatch: require('./cloudwatch'),
  CloudWatchEvents: require('./cloudwatchevents'),
  CloudWatchLogs: require('./cloudwatchlogs'),
  CodeBuild: require('./codebuild'),
  CodeCommit: require('./codecommit'),
  CodeDeploy: require('./codedeploy'),
  CodePipeline: require('./codepipeline'),
  CognitoIdentity: require('./cognitoidentity'),
  CognitoIdentityServiceProvider: require('./cognitoidentityserviceprovider'),
  CognitoSync: require('./cognitosync'),
  ConfigService: require('./configservice'),
  CUR: require('./cur'),
  DataPipeline: require('./datapipeline'),
  DeviceFarm: require('./devicefarm'),
  DirectConnect: require('./directconnect'),
  DirectoryService: require('./directoryservice'),
  Discovery: require('./discovery'),
  DMS: require('./dms'),
  DynamoDB: require('./dynamodb'),
  DynamoDBStreams: require('./dynamodbstreams'),
  EC2: require('./ec2'),
  ECR: require('./ecr'),
  ECS: require('./ecs'),
  EFS: require('./efs'),
  ElastiCache: require('./elasticache'),
  ElasticBeanstalk: require('./elasticbeanstalk'),
  ELB: require('./elb'),
  ELBv2: require('./elbv2'),
  EMR: require('./emr'),
  ES: require('./es'),
  ElasticTranscoder: require('./elastictranscoder'),
  Firehose: require('./firehose'),
  GameLift: require('./gamelift'),
  Glacier: require('./glacier'),
  Health: require('./health'),
  IAM: require('./iam'),
  ImportExport: require('./importexport'),
  Inspector: require('./inspector'),
  Iot: require('./iot'),
  IotData: require('./iotdata'),
  Kinesis: require('./kinesis'),
  KinesisAnalytics: require('./kinesisanalytics'),
  KMS: require('./kms'),
  Lambda: require('./lambda'),
  LexRuntime: require('./lexruntime'),
  Lightsail: require('./lightsail'),
  MachineLearning: require('./machinelearning'),
  MarketplaceCommerceAnalytics: require('./marketplacecommerceanalytics'),
  MarketplaceMetering: require('./marketplacemetering'),
  MTurk: require('./mturk'),
  MobileAnalytics: require('./mobileanalytics'),
  OpsWorks: require('./opsworks'),
  OpsWorksCM: require('./opsworkscm'),
  Organizations: require('./organizations'),
  Pinpoint: require('./pinpoint'),
  Polly: require('./polly'),
  RDS: require('./rds'),
  Redshift: require('./redshift'),
  Rekognition: require('./rekognition'),
  ResourceGroupsTaggingAPI: require('./resourcegroupstaggingapi'),
  Route53: require('./route53'),
  Route53Domains: require('./route53domains'),
  S3: require('./s3'),
  ServiceCatalog: require('./servicecatalog'),
  SES: require('./ses'),
  Shield: require('./shield'),
  SimpleDB: require('./simpledb'),
  SMS: require('./sms'),
  Snowball: require('./snowball'),
  SNS: require('./sns'),
  SQS: require('./sqs'),
  SSM: require('./ssm'),
  StorageGateway: require('./storagegateway'),
  StepFunctions: require('./stepfunctions'),
  STS: require('./sts'),
  Support: require('./support'),
  SWF: require('./swf'),
  XRay: require('./xray'),
  WAF: require('./waf'),
  WAFRegional: require('./wafregional'),
  WorkDocs: require('./workdocs'),
  WorkSpaces: require('./workspaces'),
  CodeStar: require('./codestar'),
  LexModelBuildingService: require('./lexmodelbuildingservice'),
  MarketplaceEntitlementService: require('./marketplaceentitlementservice'),
  Athena: require('./athena'),
  Greengrass: require('./greengrass'),
  DAX: require('./dax'),
  MigrationHub: require('./migrationhub'),
  CloudHSMV2: require('./cloudhsmv2'),
  Glue: require('./glue'),
  Mobile: require('./mobile')
};
